sockeye/lexical_constraints.py

```diff
@@ -0,0 +1,399 @@
+            if word_id != self.eos_id or self.numNeeded() == 1:
+                items.append(word_id)
+
+        # Add all constraints that aren't non-initial sequences
+        else:
+            for i, id in enumerate(self.constraints):
+                if self.met[i] is False and (i == 0 or not self.is_sequence[i - 1]):
+                    word_id = self.constraints[i]
+                    if word_id != self.eos_id or self.numNeeded() == 1:
+                        items.append(word_id)
+
+        return items
+
+    def finished(self) -> bool:
+        """
+        Return true if all the constraints have been met.
+
+        :return: True if all the constraints are met.
+        """
+        return self.numNeeded() == 0
+
+    def isValid(self, wordid) -> bool:
+        """
+        Ensures </s> is only generated when the hypothesis is completed.
+
+        :param wordid: The wordid to validate.
+        :return: True if all constraints are already met or the word ID is not the EOS id.
+        """
+        return self.finished() or wordid != self.eos_id or (self.numNeeded() == 1 and self.allowed() == [self.eos_id])
+
+    def advance(self, word_id: int) -> 'ConstrainedHypothesis':
+        """
+        Updates the constraints object based on advancing on word_id.
+        There is a complication, in that we may have started but not
+        yet completed a multi-word constraint.  We need to constraints
+        to be added as unconstrained words, so if the next word is a
+        invalid, we need to back out of marking the current phrase as
+        met constraints.
+
+        :param word_id: The word ID to advance on.
+        :return: A deep copy of the object, advanced on word_id.
+        """
+
+        obj = copy.deepcopy(self)
+
+        # First, check if we're updating a sequential constraint.
+        if obj.lastMet != -1 and obj.is_sequence[obj.lastMet] == 1:
+            if word_id == obj.constraints[obj.lastMet + 1]:
+                # Here, the word matches what we expect next in the constraint, so we update everything
+                obj.met[obj.lastMet + 1] = True
+                obj.lastMet += 1
+            else:
+                # Here, the word is not the expected next word of the constraint, so we back out of the constraint.
+                index = obj.lastMet
+                while obj.is_sequence[index]:
+                    obj.met[index] = False
+                    index -= 1
+                obj.lastMet = -1
+
+        # next, check if we can meet a single-word constraint
+        else:
+            try:
+                # build a tuple of (constraint, whether it's non-initial sequential, whether it's met)
+                l = list(zip(obj.constraints, [0] + obj.is_sequence[:-1], obj.met))
+                q = (word_id, 0, 0)
+                pos = l.index(q)
+                obj.met[pos] = True
+                obj.lastMet = pos
+
+            except:
+                pass
```

Comment:
Is it the case that any of the statements in the `try` block can fail?  Could we check something more specific to avoid the `try`/`except`/`pass` construction?

Reply:
oh, this looks like a problem. I don't remember why this was added but will fix it, thanks.